### ISSUE TITLE
[Workflow] A Transition with an array of $froms is not handled properly (Symfony 3.2)

### ISSUE THREAD

**Issue by simshaun** [symfony/symfony#20838]
$builder->addTransition(new Transition('save', 'init', 'draft'));
    $builder->addTransition(new Transition('publish', ['init', 'draft'], 'active'));

- The initial state is "init".
- Applying "save" works fine.
- Attempting to apply "publish" (after applying "save") results in an error.

This is caused by this method in `Workflow.php`:

    private function getTransitionForSubject($subject, Marking $marking, array $transitions)
    {
        foreach ($transitions as $transition) {
            foreach ($transition->getFroms() as $place) {
                if (!$marking->has($place)) {
                    continue 2;
                }
            }

            if (true !== $this->guardTransition($subject, $marking, $transition)) {
                return $transition;
            }
        }
    }

Basically, because the object is in the "draft" state after applying `save`, this method skips the `publish` transition too early, because the first element in `getFroms()` returns false for `!$marking->has($place)`.

## Edit:

I just realized this may not be a bug... but me misunderstanding how the state machine is supposed to work. Does an array in `Transition::$froms` imply that the subject must have *all* of those states, or is it supposed to mean that the subject can be *in any one* of those states?

If this isn't a bug , the correct solution would be this, right?

    $builder->addTransition(new Transition('save', 'init', 'draft'));
    $builder->addTransition(new Transition('publish', 'init', 'active'));
    $builder->addTransition(new Transition('publish', 'draft', 'active'));

**Comment by kmark** [comment]
Mentioned this in the PR thread, but again, thanks for your help in addressing this bug.

**Comment by simshaun** [comment]
@kmark brought up a good point in IRC. Both scenarios produce (practically) the same graph.

If this issue isn't actually an issue, I suppose there is a bug in the GraphvizDumper?

**Scenario 1:**

    $builder->addTransition(new Transition('save', 'init', 'draft'));
    $builder->addTransition(new Transition('publish', ['init', 'draft'], 'active'));

![image](https://cloud.githubusercontent.com/assets/379519/21036570/c3648d60-bd94-11e6-9b52-1f2e22035287.png)


**Scenario 2:**

    $builder->addTransition(new Transition('save', 'init', 'draft'));
    $builder->addTransition(new Transition('publish', 'init', 'active'));
    $builder->addTransition(new Transition('publish', 'draft', 'active'));

![image](https://cloud.githubusercontent.com/assets/379519/21036575/d561cdc0-bd94-11e6-8604-453776d07fe1.png)

**Comment by simshaun** [comment]
Look at the PullRequest example given in the docs @ https://symfony.com/doc/master/workflow/state-machines.html.

It shows the **update** transition being valid from `[coding, travis, review]`.

I take that to mean:

> A `PullRequest` may be *update*d if it's in coding, travis, OR review state.

So maybe this bug report actually is valid? lol

**Comment by HeahDude** [comment]
Hi @simshaun 
>I just realized this may not be a bug... but me misunderstanding how the state machine is supposed to work. Does an array in Transition::$froms imply that the subject must have all of those states, or is it supposed to mean that the subject can be in any one of those states?

A workflow implies to be in all "Froms" to get to the "Tos".

>If this isn't a bug , the correct solution would be this, right?

>$builder->addTransition(new Transition('save', 'init', 'draft'));
>$builder->addTransition(new Transition('publish', 'init', 'active'));
>$builder->addTransition(new Transition('publish', 'draft', 'active'));

This is indeed the solution, but you're now defining a state machine which is a subset of workflow and implies transitions with one "From" to one "To".

So to ease the implementation of state machines it has been allowed to have transitions with the same name to handle many "Froms" to one "To".

**Comment by HeahDude** [comment]
Let's close here as well.

**Comment by simshaun** [comment]
Did you see my last comment here? The docs are contradictory to your statement, no?

**Comment by HeahDude** [comment]
Please open an issue in the docs if you feel something is too misleading. Thanks!

**Comment by stof** [comment]
@simshaun the graphviz dumper indeed does not handle properly workflows where transition names are not unique (due to the fact that uniqueness was a requirement at the time we implemented the dumper based on this assumption, and the requirement was removed later)

**Comment by dfu-io** [comment]
Hi,
I'm having the same problem when I try to define more than one origin.

For example, I've got the next available places:
[parked, started, moving, crashed]

And the next available transitions:
start: parked => started
move: started => moving
park: started, moving => parked
crash: moving => crashed

I've tried with individual transitions (simplified):

```php
$builder->addTransition(new Transition('start', 'parked', 'started'));
$builder->addTransition(new Transition('move', 'started', 'moving'));
$builder->addTransition(new Transition('park', 'started', 'parked'));
$builder->addTransition(new Transition('park', 'moving', 'parked'));
$builder->addTransition(new Transition('crash', 'moving', 'crashed'));

$definition = $builder->build();
$stateMachine = new StateMachine($definition, new SingleStateMarkingStore('state'));
$registry->add($stateMachine, Car::class)
```
And with arrays:
```php
$builder->addTransition(new Transition('start', 'parked', 'started'));
$builder->addTransition(new Transition('move', 'started', 'moving'));
$builder->addTransition(new Transition('park', ['started', 'moving'], 'parked'));
$builder->addTransition(new Transition('crash', 'moving', 'crashed'));

$definition = $builder->build();
$stateMachine = new StateMachine($definition, new SingleStateMarkingStore('state'));
$registry->add($stateMachine, Car::class)
```

If I tray with individual transitions, only works the first. If a try with arrays, doesn't works never.

Any idea? I'm using the last version of workflow: "^3.2" (In this case used as a component without Symfony)
